src/bms/parse/token_processor.rs

```diff
@@ -293,6 +294,80 @@ pub(crate) fn full_preset<T: KeyLayoutMapper, R: Rng>(
     )
 }
 
+pub(crate) fn relax_tokens_default<'a>(tokens: &mut TokenStream<'a>) {
+    for twr in tokens.tokens.iter_mut() {
+        match twr.content_mut() {
+            Token::Header { name, args } => {
+                let n_ref = name.as_ref();
+                let a_ref = args.as_ref();
+                let mut new_name: Option<String> = None;
+                let mut new_args: Option<String> = None;
+
+                if n_ref.eq_ignore_ascii_case("RONDAM") {
+                    new_name = Some("RANDOM".to_string());
+                } else if n_ref.eq_ignore_ascii_case("END")
+                    && a_ref.trim().eq_ignore_ascii_case("IF")
+                {
+                    new_name = Some("ENDIF".to_string());
+                    new_args = Some(String::new());
+                } else if a_ref.is_empty()
+                    && let Some((kw, rest_trim)) = ["RANDOM", "IF"]
+                        .iter()
+                        .find_map(|kw| n_ref.strip_prefix_ignore_case(kw).map(|r| (*kw, r.trim())))
+                {
+                    let digits = if rest_trim.starts_with('[') && rest_trim.ends_with(']') {
+                        &rest_trim[1..rest_trim.len() - 1]
+                    } else {
+                        rest_trim
+                    };
+                    if !digits.is_empty() && digits.chars().all(|c| c.is_ascii_digit()) {
+                        new_name = Some(kw.to_string());
+                        new_args = Some(digits.to_string());
+                    }
+                }
+
+                if let Some(nn) = new_name {
+                    *name = nn.into();
+                }
+                if let Some(na) = new_args {
+                    *args = na.into();
+                }
+            }
+            Token::NotACommand(line) => {
+                if line.trim() == "＃ENDIF" {
+                    *twr.content_mut() = Token::Header {
+                        name: "ENDIF".to_string().into(),
+                        args: String::new().into(),
+                    };
+                }
+            }
+            Token::Message { .. } => {}
+        }
+    }
+}
+
+/// Transforms lex tokens before semantic parsing.
+///
+/// Implementations can rewrite headers, normalize arguments, or fix common typos.
+/// Called before semantic token processors consume the stream.
+pub trait TokenModifier {
+    /// Applies in-place modifications to the provided token stream.
+    ///
+    /// Implementations should be deterministic and avoid altering source ranges.
+    fn modify(&self, tokens: &mut TokenStream<'_>);
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash)]
+/// A relaxed modifier that normalizes common BMS typos and formats,
+/// delegating to `rewrite_relaxed_tokens`.
+pub struct DefaultTokenRelaxer;
+
+impl TokenModifier for DefaultTokenRelaxer {
+    fn modify(&self, tokens: &mut TokenStream<'_>) {
+        relax_tokens_default(tokens);
+    }
+}
```

Comment:
Alike as `TokenProcessor` for `SequentialProcessor`, you can add a sequential join type for two of `TokenModifier`s. If you do this, `ParseConfig` can avoid to use `Vec<Box<dyn TokenModifier>>` and available the static dispatch with type parameters.